File: app/schemas/horas_extra_solicitud.py
"""Schemas para solicitudes de horas extra (supervisor)."""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

from app.schemas.horas_extra_aprobacion import HorasExtraEstadoConsolidado

HorasExtraTipoSolicitud = Literal["planeado", "espontaneo"]
HorasExtraEstadoSolicitud = Literal[
    "borrador", "pendiente", "aprobado", "rechazado", "cancelado"
]


class HorasExtraEmpleadoOption(BaseModel):
    id: int
    no_empleado: int
    nombre: str
    centrocosto_id: Optional[int] = None
    area_id: Optional[int] = None
    subarea_id: Optional[int] = None
    area_descripcion: Optional[str] = None
    centrocosto_descripcion: Optional[str] = None
    turno: Optional[str] = None


class HorasExtraSolicitudOpcionesResponse(BaseModel):
    empleados: list[HorasExtraEmpleadoOption]
    semana_actual: int = Field(ge=1, le=53)


class HorasExtraDetalleCreate(BaseModel):
    empleado_id: int
    lunes: Decimal = Field(default=Decimal("0"), ge=0)
    martes: Decimal = Field(default=Decimal("0"), ge=0)
    miercoles: Decimal = Field(default=Decimal("0"), ge=0)
    jueves: Decimal = Field(default=Decimal("0"), ge=0)
    viernes: Decimal = Field(default=Decimal("0"), ge=0)
    sabado: Decimal = Field(default=Decimal("0"), ge=0)
    domingo: Decimal = Field(default=Decimal("0"), ge=0)


class HorasExtraSolicitudCreate(BaseModel):
    semana: int = Field(ge=1, le=53)
    tipo: HorasExtraTipoSolicitud
    motivo: str = Field(min_length=1, max_length=500)
    empleados: list[HorasExtraDetalleCreate] = Field(min_length=1)

    @field_validator("motivo")
    @classmethod
    def strip_motivo(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("El motivo es obligatorio.")
        return stripped


class HorasExtraDetalleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    empleado_id: int
    no_empleado: int
    nombre_empleado: str
    lunes: Decimal
    martes: Decimal
    miercoles: Decimal
    jueves: Decimal
    viernes: Decimal
    sabado: Decimal
    domingo: Decimal
    total_horas: Decimal

    @field_serializer(
        "lunes",
        "martes",
        "miercoles",
        "jueves",
        "viernes",
        "sabado",
        "domingo",
        "total_horas",
    )
    def _serialize_horas(self, value: Decimal) -> float:
        return float(value)


class HorasExtraSolicitudResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    fecha_solicitud: date
    semana: int
    semana_inicio: date
    tipo: HorasExtraTipoSolicitud
    area_id: int
    area_descripcion: str
    subarea_id: int
    subarea_descripcion: str
    centrocosto_id: int
    centrocosto_descripcion: str
    motivo_id: int
    motivo_descripcion: str
    comentarios: Optional[str] = None
    estado: HorasExtraEstadoSolicitud
    estado_consolidado: HorasExtraEstadoConsolidado
    total_horas_general: Decimal
    total_empleados: int
    created_at: datetime
    detalle: list[HorasExtraDetalleResponse]

    @field_serializer("total_horas_general")
    def _serialize_total_general(self, value: Decimal) -> float:
        return float(value)


class HorasExtraSolicitudListItem(BaseModel):
    id: int
    fecha_solicitud: date
    semana: int
    semana_inicio: date
    area_descripcion: str
    tipo: HorasExtraTipoSolicitud
    total_horas_general: Decimal
    estado: HorasExtraEstadoSolicitud
    estado_consolidado: HorasExtraEstadoConsolidado
    created_at: datetime

    @field_serializer("total_horas_general")
    def _serialize_total_general(self, value: Decimal) -> float:
        return float(value)


class HorasExtraSolicitudListResponse(BaseModel):
    items: list[HorasExtraSolicitudListItem]
    total: int
    page: int
    page_size: int


class HorasExtraSolicitudEstadisticasResponse(BaseModel):
    total_solicitudes: int = Field(ge=0)
    pendientes: int = Field(ge=0)
    aprobadas: int = Field(ge=0)
    total_horas: Decimal = Field(ge=0)

    @field_serializer("total_horas")
    def _serialize_total_horas(self, value: Decimal) -> float:
        return float(value)

File: frontend/src/api/horasExtraSolicitud.ts
import { fetchWithAuth } from "./http.ts";

async function readErrorDetail(res: Response): Promise<string> {
  const raw = await res.text();
  try {
    const j = JSON.parse(raw) as { detail?: unknown };
    if (typeof j.detail === "string") return j.detail;
  } catch {
    /* ignore */
  }
  return raw || res.statusText || "Error";
}

export type HorasExtraTipoSolicitud = "planeado" | "espontaneo";
export type HorasExtraEstadoSolicitud =
  | "borrador"
  | "pendiente"
  | "aprobado"
  | "rechazado"
  | "cancelado";

export type HorasExtraEmpleadoOption = {
  id: number;
  no_empleado: string;
  nombre: string;
  centrocosto_id: number | null;
  area_id: number | null;
  subarea_id: number | null;
};

export type HorasExtraSolicitudOpciones = {
  empleados: HorasExtraEmpleadoOption[];
  semana_actual: number;
};

export type HorasExtraDetalleCreate = {
  empleado_id: number;
  lunes: number;
  martes: number;
  miercoles: number;
  jueves: number;
  viernes: number;
  sabado: number;
  domingo: number;
};

export type HorasExtraSolicitudCreate = {
  fecha_solicitud: string;
  semana: number;
  tipo: HorasExtraTipoSolicitud;
  motivo: string;
  empleados: HorasExtraDetalleCreate[];
};

export type HorasExtraDetalleResponse = HorasExtraDetalleCreate & {
  id: number;
  no_empleado: string;
  nombre_empleado: string;
  total_horas: number;
};

export type HorasExtraSolicitudResponse = {
  id: number;
  fecha_solicitud: string;
  semana: number;
  semana_inicio: string;
  tipo: HorasExtraTipoSolicitud;
  departamento_id: number;
  departamento_nombre: string;
  area_id: number;
  area_descripcion: string;
  subarea_id: number;
  subarea_descripcion: string;
  centrocosto_id: number;
  centrocosto_descripcion: string;
  motivo_id: number;
  motivo_descripcion: string;
  comentarios: string | null;
  estado: HorasExtraEstadoSolicitud;
  total_horas_general: number;
  total_empleados: number;
  created_at: string;
  detalle: HorasExtraDetalleResponse[];
};

export type HorasExtraSolicitudListItem = {
  id: number;
  fecha_solicitud: string;
  semana: number;
  semana_inicio: string;
  departamento_nombre: string;
  area_descripcion: string;
  tipo: HorasExtraTipoSolicitud;
  total_horas_general: number;
  estado: HorasExtraEstadoSolicitud;
  created_at: string;
};

export type HorasExtraSolicitudListResponse = {
  items: HorasExtraSolicitudListItem[];
  total: number;
  page: number;
  page_size: number;
};

export type HorasExtraSolicitudFetchError = {
  status: number;
  detail: string;
};

export async function getHorasExtraSolicitudOpciones(): Promise<HorasExtraSolicitudOpciones> {
  const res = await fetchWithAuth("/api/v1/horas-extra/solicitudes/opciones");
  if (!res.ok) {
    const err: HorasExtraSolicitudFetchError = {
      status: res.status,
      detail: await readErrorDetail(res),
    };
    throw err;
  }
  return (await res.json()) as HorasExtraSolicitudOpciones;
}

export async function getHorasExtraSolicitudes(
  page = 1,
  pageSize = 10,
): Promise<HorasExtraSolicitudListResponse> {
  const sp = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });
  const res = await fetchWithAuth(`/api/v1/horas-extra/solicitudes?${sp.toString()}`);
  if (!res.ok) {
    const err: HorasExtraSolicitudFetchError = {
      status: res.status,
      detail: await readErrorDetail(res),
    };
    throw err;
  }
  return (await res.json()) as HorasExtraSolicitudListResponse;
}

export async function createHorasExtraSolicitud(
  body: HorasExtraSolicitudCreate,
): Promise<HorasExtraSolicitudResponse> {
  const res = await fetchWithAuth("/api/v1/horas-extra/solicitudes", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const err: HorasExtraSolicitudFetchError = {
      status: res.status,
      detail: await readErrorDetail(res),
    };
    throw err;
  }
  return (await res.json()) as HorasExtraSolicitudResponse;
}

export async function getHorasExtraSolicitudDetalle(
  id: number,
): Promise<HorasExtraSolicitudResponse> {
  const res = await fetchWithAuth(`/api/v1/horas-extra/solicitudes/${id}`);
  if (!res.ok) {
    const err: HorasExtraSolicitudFetchError = {
      status: res.status,
      detail: await readErrorDetail(res),
    };
    throw err;
  }
  return (await res.json()) as HorasExtraSolicitudResponse;
}
